### PR TITLE
Allow ILM to transition to implicit cached steps

### LINKED ISSUE #91749: [ILM] Error moving to the phase `complete` step when phase was deleted
### Elasticsearch Version

7.17, 8.6, 8.5

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Darwin

### Problem Description

Some ILM steps are implicit ie. injected (`migrate` action) or the policy/phase `complete` step.

ILM tries to honour the cached ILM phase however, these implicit steps are not **seen** as cached so might still end up with errors like:
```
step [{"phase":"warm","action":"complete","name":"complete"}] for index [index] with policy [my-policy] does not exist
```
or
```
step [{"phase":"warm","action":"migrate","name":"check-migration"}] for index [index] with policy [my-policy] does not exist
```

The problem lies [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java#L86) where we only look at the the explicitly declared actions in the cached phase

### Steps to Reproduce

```    
public void testValidateTransitionToCachedStepWhenMissingPhaseFromPolicy() {
        // we'll test the case when the warm phase was deleted and the next step is the phase complete one

        LifecycleExecutionState.Builder executionState = LifecycleExecutionState.builder()
            .setPhase("warm")
            .setAction("migrate")
            .setStep("check-migration")
            .setPhaseDefinition("""
                {
                        "policy" : "my-policy",
                        "phase_definition" : {
                          "min_age" : "20m",
                          "actions" : {
                            "set_priority" : {
                              "priority" : 150
                            }
                          }
                        },
                        "version" : 1,
                        "modified_date_in_millis" : 1578521007076
                      }""");

        IndexMetadata meta = buildIndexMetadata("my-policy", executionState);

        try (Client client = new NoOpClient(getTestName())) {
            Step.StepKey currentStepKey = new Step.StepKey("warm", MigrateAction.NAME, DataTierMigrationRoutedStep.NAME);
            Step.StepKey nextStepKey = new Step.StepKey("warm", PhaseCompleteStep.NAME, PhaseCompleteStep.NAME);

            Step.StepKey waitForRolloverStepKey = new Step.StepKey("hot", RolloverAction.NAME, WaitForRolloverReadyStep.NAME);
            Step.StepKey rolloverStepKey = new Step.StepKey("hot", RolloverAction.NAME, RolloverStep.NAME);
            Step waitForRolloverReadyStep = new WaitForRolloverReadyStep(
                waitForRolloverStepKey,
                rolloverStepKey,
                client,
                null,
                null,
                null,
                1L,
                null,
                null,
                null,
                null,
                null,
                null
            );

            try {
                IndexLifecycleTransition.validateTransition(
                    meta,
                    currentStepKey,
                    nextStepKey,
                    createOneStepPolicyStepRegistry("my-policy", waitForRolloverReadyStep)
                );
            } catch (Exception e) {
                logger.error(e.getMessage(), e);
                fail("validateTransition should not throw exception on valid transitions");
            }
        }
    }

    public void testValidateTransitionToInjectedMissingStep() {
        // we'll test the case when the warm phase was deleted and the next step is an injected one

        LifecycleExecutionState.Builder executionState = LifecycleExecutionState.builder()
            .setPhase("warm")
            .setAction("migrate")
            .setStep("migrate")
            .setPhaseDefinition("""
                {
                        "policy" : "my-policy",
                        "phase_definition" : {
                          "min_age" : "20m",
                          "actions" : {
                            "set_priority" : {
                              "priority" : 150
                            }
                          }
                        },
                        "version" : 1,
                        "modified_date_in_millis" : 1578521007076
                      }""");

        IndexMetadata meta = buildIndexMetadata("my-policy", executionState);

        try (Client client = new NoOpClient(getTestName())) {
            Step.StepKey currentStepKey = new Step.StepKey("warm", MigrateAction.NAME, MigrateAction.NAME);
            Step.StepKey nextStepKey = new Step.StepKey("warm", MigrateAction.NAME, DataTierMigrationRoutedStep.NAME);

            Step.StepKey waitForRolloverStepKey = new Step.StepKey("hot", RolloverAction.NAME, WaitForRolloverReadyStep.NAME);
            Step.StepKey rolloverStepKey = new Step.StepKey("hot", RolloverAction.NAME, RolloverStep.NAME);
            Step waitForRolloverReadyStep = new WaitForRolloverReadyStep(
                waitForRolloverStepKey,
                rolloverStepKey,
                client,
                null,
                null,
                null,
                1L,
                null,
                null,
                null,
                null,
                null,
                null
            );

            try {
                IndexLifecycleTransition.validateTransition(
                    meta,
                    currentStepKey,
                    nextStepKey,
                    createOneStepPolicyStepRegistry("my-policy", waitForRolloverReadyStep)
                );
            } catch (Exception e) {
                logger.error(e.getMessage(), e);
                fail("validateTransition should not throw exception on valid transitions");
            }
        }
    }
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by andreidan**
ILM tries to honour the cached phase however, some steps are implicit (e.g. injected actions or the terminal policy/phase step) Currently, ILM would throw an exception if the currently cached phase was removed:
```
step [{"phase":"warm","action":"complete","name":"complete"}] for index [index]
with policy [my-policy] does not exist
```
ILM would currently also throw an exception if the next step was an implicit one and the phase was removed from the underlying policy e.g. if the index is on the `migrate` step and looking to transition to the `check-migration` step, whilt the `warm` phase doesn't exist in the policy anymore
```
step [{"phase":"warm","action":"migrate","name":"check-migration"}] for index
[index] with policy [my-policy] does not exist
```

This fixes these scenarios by enhancing the
`PolicyStepsRegistry#parseStepKeysFromPhase` method to compute all the steps in the phase (including all implicit steps)

Fixes #91749

### CODE REVIEW

**Comment by andreidan** [comment]
Note: there might be a slight performance penalty here due to `LifecyclePolicy#toSteps`( https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecyclePolicy.java#L215 )
now executing on every transition

**Comment by andreidan** [comment]
Potentially superseding https://github.com/elastic/elasticsearch/pull/91754 (pending discussion on the potential performance drop)

**Comment by dakrone** [comment]
@joegallo I think you introduced the `PhaseCacheManagement` piece, which was to increase the performance of this specific method if I recall correctly. Do you have thoughts on how this will affect performance?

**Comment by andreidan** [comment]
@dakrone I introduced the `PhaseCacheManagement` to handle the phase manipulations we perform when migrating a deployment to data tiers (ie. we potentially remove the `allocate` action from phases). The reason this class exists is for encapsulating the interaction and management of the cached phase.

The change I propose here eludes the `PhaseCacheManagement` to keep the change as isolated as possible (ie. code affects only the transition validation to the next step) - hence the code is in `PolicyStepsRegistry`. 

However, we'll continue to disregard implicit steps when parsing the cached phase (as we do today) when it comes to establishing if the current cached phase is safely updateable when updating a policy (ie. `PhaseCacheManagement#isIndexPhaseDefinitionUpdatable` will continue to work as it is today)

@joegallo what do you think ?

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by joegallo** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java [hunk 1]
```diff
@@ -245,11 +246,31 @@ public Step.StepKey getFirstStepForPhaseAndAction(ClusterState state, Index inde
...
+     * This method returns **all** the steps that are part of the phase definition including the implicit steps.
+     *
      * Returns null if there's a parsing error.
      */
     @Nullable
-    public Set<Step.StepKey> parseStepKeysFromPhase(String phaseDef, String currentPhase) {
-        return PhaseCacheManagement.readStepKeys(xContentRegistry, client, phaseDef, currentPhase, licenseState);
+    public Set<Step.StepKey> parseStepKeysFromPhase(String policy, String currentPhase, String phaseDef) {
+        try {
+            String phaseDefNonNull = Optional.ofNullable(phaseDef).orElse(InitializePolicyContextStep.INITIALIZATION_PHASE);
```
Comment (joegallo) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java:259]: Maybe `Objects.requireNonNullElse(phaseDef, InitializePolicyContextStep.INITIALIZATION_PHASE)`?

In the same neighborhood, does it make sense to log the `null` in the `logger.trace` below or would it be better to drop `phaseDefNonNull` and just keep everything in terms of `phaseDef`?

**Comment by joegallo** [comment]
I'm worried that this is going to blow up our performance for the many shards use case. `parseStepKeysFromPhase` is called from `IndexLifecycleTransition.validateTransition` and that's called by `IndexLifecycleTransition.moveClusterStateToStep`.

That last one is part of the main ILM engine via `ExecuteStepsUpdateTask` and `MoveToNextStepUpdateTask`.

Given the number of backports this is going to generate, I think it's best if we characterize the slowdown that we'll see in the many shards case (if any!) before merging it.

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java [hunk 2]
```diff
@@ -632,6 +634,126 @@ public void testValidateTransitionToCachedStepMissingFromPolicy() {
...
 
+    public void testValidateTransitionToCachedStepWhenMissingPhaseFromPolicy() {
+        // we'll test the case when the warm phase was deleted and the next step is the phase complete one
+
+        LifecycleExecutionState.Builder executionState = LifecycleExecutionState.builder()
+            .setPhase("warm")
+            .setAction("migrate")
+            .setStep("check-migration")
+            .setPhaseDefinition("""
+                {
```
Comment (joegallo) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java:645]: Gahhhh! Spotless!!! There's gotta be something we can do to make this prettier.

Comment (joegallo) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransitionTests.java:645]: #92713 and 822adffbbd7d423ad

**Comment by joegallo** [comment]
I've assigned this to myself -- I want to do a little more looking at `parseStepKeysFromPhase` and `parseStepsFromPhase`. I'm not in love with the Stringly-typed `phaseDef` argument we're using for both. It can be null, a phase name like 'new' or 'complete' or a full blown json String that we parse. I'd like to see if I can tighten that up a bit.

**Comment by andreidan** [comment]
@joegallo do you have an update here? Could we try to benchmark this approach and see if it's feasible?

**Comment by andreidan** [comment]
@elasticmachine update branch

**Review by joegallo — APPROVED** [review]

**Comment by joegallo** [comment]
The plan is to merge this and give the benchmarks a little while to show an effect, then consult with Armin later this week to determine the path forward (if there's no effect, then that would imply a short meeting 😉).

**Pull request merged**